core/src/main/java/org/apache/iceberg/io/IOUtil.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+public class IOUtil {
+  // not meant to be instantiated
+  private IOUtil() {
+  }
+
+  /**
+   * Reads into a buffer from a stream, making multiple read calls if necessary.
+   *
+   * @param stream an InputStream to read from
+   * @param bytes a buffer to write into
+   * @param offset starting offset in the buffer for the data
+   * @param length length of bytes to copy from the input stream to the buffer
+   * @throws IOException if there is an error while reading
+   */
+  @SuppressWarnings("checkstyle:InnerAssignment")
+  public static void readFully(InputStream stream, byte[] bytes, int offset, int length) throws IOException {
+    if (readToEnd(stream, bytes, offset, length) != length) {
+      throw new IOException("End of stream reached before completing read");
+    }
+  }
+
+  /**
+   * Reads into a buffer from a stream, making multiple read calls if necessary
+   * returning the number of bytes read until end of stream.
+   *
+   * @param stream an InputStream to read from
+   * @param bytes a buffer to write into
+   * @param offset starting offset in the buffer for the data
+   * @param length length of bytes to copy from the input stream to the buffer
+   * @throws IOException if there is an error while reading
+   */
+  @SuppressWarnings("checkstyle:InnerAssignment")
+  public static int readToEnd(InputStream stream, byte[] bytes, int offset, int length) throws IOException {
+    int pos = offset;
+    int bytesRead = 0;
+
+    while ((length - bytesRead) > 0 && (bytesRead = stream.read(bytes, pos, length - bytesRead)) > 0) {
+      pos += bytesRead;
+    }
```

Comment:
Is this loop correct ? 

If we have a stream with 1000 bytes, and the offset = 0,  length = 100,  assume we will read 10 bytes for each `stream.read(...)`. Then:

1. For the first read,  everything works fine, and the `pos` advanced to 10; 
2. For the second read,    it will try to read by the `stream.read(bytes, 10, 100-10) `.  Then the `pos` advanced to  20; 
3. For the third read, it will try to read by the `stream.read(bytes, 20, 100-10)`.  In fact,  we only need to read the remaining 80 bytes but we are passing the `100-10=90` bytes ?  That should be a problem....

Reply:
This was my bad. I found the wrong `readFully` implementation and didn't update it correctly.

I'm opening a PR against this branch with a better implementation and thorough tests for the method.

Reply:
Opened https://github.com/danielcweeks/iceberg/pull/1 to improve this. @danielcweeks can you take a look?

Reply:
Great catch @openinx and thanks for the update @rdblue.  I merged that implementation in.